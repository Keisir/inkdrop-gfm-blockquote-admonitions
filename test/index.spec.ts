// Copyright (C) myl7
// SPDX-License-Identifier: Apache-2.0

import { selectAll, selectOne } from 'css-select'
import { parseDocument } from 'htmlparser2'
import rehypeStringify from 'rehype-stringify'
import { remark } from 'remark'
import remarkParse from 'remark-parse'
import remarkRehype from 'remark-rehype'
import { describe, expect, it } from 'vitest'
import plugin from '../src/index.js'

function defineCase(
  name: string,
  options: {
    input: string
    assertions: (html: string) => Promise<void> | void
  }
) {
  it(name, async function () {
    const processor = remark()
      .use(remarkParse)
      // To make unified happy, since it does not allow (for types) to pass an undefined as config.
      // This test code can also handle the difference between `.use(plugin)` and `.use(plugin, {})`.
      .use(plugin)
      .use(remarkRehype)
      .use(rehypeStringify)

    const html = String(await processor.process(options.input))
    await options.assertions(html)

    expect(html).toMatchSnapshot()
  })
}

describe('GitHub beta blockquote-based admonitions with titles like [!NOTE]', function () {
  defineCase('should transform', {
    input: `\
# Admonitions
> [!NOTE]
> test
`,
    assertions(html) {
      const elem = selectOne(
        'div.gfm-alert > p.gfm-alert-title:first-child',
        parseDocument(html)
      )
      expect(elem).to.have.nested.property('firstChild.data', 'Note')
    }
  })

  defineCase('should not transform when single line', {
    input: `\
# Admonitions
> [!NOTE] test
`,
    assertions(html) {
      const elem = selectOne('div.gfm-alert', parseDocument(html))
      expect(elem).to.have.be.null
    }
  })

  defineCase('should transform with nested ones', {
    input: `\
# Admonitions
> [!NOTE]
> test
>
> > [!NOTE]
> > test
> >
> > > [!WARNING]
> > > test
`,
    assertions(html) {
      const elem = selectOne(
        'div.gfm-alert > div.gfm-alert > div.gfm-alert > p.gfm-alert-title:first-child',
        parseDocument(html)
      )
      expect(elem).to.have.nested.property('firstChild.data', 'Warning')
    }
  })

  defineCase('should transform with unordered lists from issue #4', {
    input: `\
# Admonitions
> [!NOTE]
> - Here you go
> - Here you go again
> - Here you go one more time
`,
    assertions(html) {
      const elem = selectOne(
        'div.gfm-alert > p.gfm-alert-title:first-child',
        parseDocument(html)
      )
      expect(elem).to.have.nested.property('firstChild.data', 'Note')
    }
  })

  defineCase('should transform with ordered lists from issue #4', {
    input: `\
# Admonitions
> [!NOTE]
> 1. Here you go
> 2. Here you go again
> 3. Here you go one more time
`,
    assertions(html) {
      const elem = selectOne(
        'div.gfm-alert > p.gfm-alert-title:first-child',
        parseDocument(html)
      )
      expect(elem).to.have.nested.property('firstChild.data', 'Note')
    }
  })

  defineCase(
    'should not transform when title is not in form [!NOTE] but legacy **Note**',
    {
      input: `\
# Admonitions
> **Note**
> test
`,
      assertions(html) {
        const elem = selectOne('div.gfm-alert', parseDocument(html))
        expect(elem).to.be.null
      }
    }
  )

  defineCase(
    'should transform with title with trailing whitespaces to be trimmed',
    {
      input: `\
# Admonitions
> [!NOTE] \r\t\v\

> test
`,
      assertions(html) {
        const elem = selectOne(
          'div.gfm-alert > p.gfm-alert-title:first-child',
          parseDocument(html)
        )
        expect(elem).to.have.nested.property('firstChild.data', 'Note')
      }
    }
  )

  defineCase(
    'should transform GitHub example (until 2024-01-02) with default config',
    {
      input: `\
# Admonitions
> [!NOTE]
> Highlights information that users should take into account, even when skimming.

> [!TIP]
> Optional information to help a user be more successful.

> [!IMPORTANT]
> Crucial information necessary for users to succeed.

> [!WARNING]
> Critical content demanding immediate user attention due to potential risks.

> [!CAUTION]
> Negative potential consequences of an action.
`,
      assertions(html) {
        const elems = selectAll(
          'div.gfm-alert > p.gfm-alert-title:first-child',
          parseDocument(html)
        )
        expect(elems).to.have.lengthOf(5)
        expect(elems[0]).to.have.nested.property('firstChild.data', 'Note')
        expect(elems[1]).to.have.nested.property('firstChild.data', 'Tip')
        expect(elems[2]).to.have.nested.property('firstChild.data', 'Important')
        expect(elems[3]).to.have.nested.property('firstChild.data', 'Warning')
        expect(elems[4]).to.have.nested.property('firstChild.data', 'Caution')
      }
    }
  )

  defineCase(
    'should transform when title has 2 trailing spaces from issue #12',
    {
      input:
        '# Admonitions\n' +
        // So that the title isn't put inline with the forecoming text,
        // when no GFM admonitions are available.
        // These 2 spaces with the newline are transformed into an inline break.
        // ----------vv
        '> [!WARNING]  \n' +
        '> Critical content demanding immediate user attention due to potential risks.\n',
      assertions(html) {
        const elem = selectOne(
          'div.gfm-alert > p.gfm-alert-title:first-child',
          parseDocument(html)
        )
        expect(elem).to.have.nested.property('firstChild.data', 'Warning')
      }
    }
  )

  defineCase('should transform a plain blockquote #13', {
    input: '> blockquote\n',
    assertions(html) {
      const elem = selectOne('blockquote > p', parseDocument(html))
      expect(elem).to.have.nested.property('firstChild.data', 'blockquote')
    }
  })

  defineCase('should transform a plain blockquote #14', {
    input: '> > line 1\n> > line 2\n> > line 3',
    assertions(html) {
      const elem = selectOne('blockquote > p', parseDocument(html))
      expect(elem).to.have.nested.property(
        'firstChild.data',
        'line 1\nline 2\nline 3'
      )
    }
  })
})

describe('verify default behavior with plugin enabled', function () {
  defineCase('should transform a plain blockquote with line breaks (\\)', {
    input: `\
> Each time you increase the amount of code, your software grows exponentially more complicated.\\
> -- DHH, Getting Real
`,
    assertions(html) {
      const elem = selectAll('blockquote > p > br', parseDocument(html))
      expect(elem).to.lengthOf(1)
    }
  })

  defineCase(
    'should transform a plain blockquote with line breaks ("  " - note the two spaces)',
    {
      input: `\
> Each time you increase the amount of code, your software grows exponentially more complicated.  
> -- DHH, Getting Real
`,
      assertions(html) {
        const elem = selectAll('blockquote > p > br', parseDocument(html))
        expect(elem).to.lengthOf(1)
      }
    }
  )

  defineCase('should transform a plain blockquote with line breaks (\\) 2x', {
    input: `\
> Each time you increase the amount of code, your software grows exponentially more complicated.\\
> -- DHH, Getting Real\\
> -- DHH, Getting Real
`,
    assertions(html) {
      const elem = selectAll('blockquote > p > br', parseDocument(html))
      expect(elem).to.lengthOf(2)
    }
  })
})
