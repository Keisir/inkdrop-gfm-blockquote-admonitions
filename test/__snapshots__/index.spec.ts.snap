// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GitHub beta blockquote-based admonitions with titles like [!NOTE] > should not transform when single line 1`] = `
"<h1>Admonitions</h1>
<blockquote>
<p>[!NOTE] test</p>
</blockquote>"
`;

exports[`GitHub beta blockquote-based admonitions with titles like [!NOTE] > should not transform when title is not in form [!NOTE] but legacy **Note** 1`] = `
"<h1>Admonitions</h1>
<blockquote>
<p><strong>Note</strong>
test</p>
</blockquote>"
`;

exports[`GitHub beta blockquote-based admonitions with titles like [!NOTE] > should transform 1`] = `
"<h1>Admonitions</h1>
<div class="gfm-alert gfm-alert-type-note">
<p class="gfm-alert-title">Note</p>
<p>test</p>
</div>"
`;

exports[`GitHub beta blockquote-based admonitions with titles like [!NOTE] > should transform GitHub example (until 2024-01-02) with default config 1`] = `
"<h1>Admonitions</h1>
<div class="gfm-alert gfm-alert-type-note">
<p class="gfm-alert-title">Note</p>
<p>Highlights information that users should take into account, even when skimming.</p>
</div>
<div class="gfm-alert gfm-alert-type-tip">
<p class="gfm-alert-title">Tip</p>
<p>Optional information to help a user be more successful.</p>
</div>
<div class="gfm-alert gfm-alert-type-important">
<p class="gfm-alert-title">Important</p>
<p>Crucial information necessary for users to succeed.</p>
</div>
<div class="gfm-alert gfm-alert-type-warning">
<p class="gfm-alert-title">Warning</p>
<p>Critical content demanding immediate user attention due to potential risks.</p>
</div>
<div class="gfm-alert gfm-alert-type-caution">
<p class="gfm-alert-title">Caution</p>
<p>Negative potential consequences of an action.</p>
</div>"
`;

exports[`GitHub beta blockquote-based admonitions with titles like [!NOTE] > should transform a plain blockquote #13 1`] = `
"<blockquote>
<p>blockquote</p>
</blockquote>"
`;

exports[`GitHub beta blockquote-based admonitions with titles like [!NOTE] > should transform a plain blockquote #14 1`] = `
"<blockquote>
<blockquote>
<p>line 1
line 2
line 3</p>
</blockquote>
</blockquote>"
`;

exports[`GitHub beta blockquote-based admonitions with titles like [!NOTE] > should transform when title has 2 trailing spaces from issue #12 1`] = `
"<h1>Admonitions</h1>
<div class="gfm-alert gfm-alert-type-warning">
<p class="gfm-alert-title">Warning</p>
<p>Critical content demanding immediate user attention due to potential risks.</p>
</div>"
`;

exports[`GitHub beta blockquote-based admonitions with titles like [!NOTE] > should transform with nested ones 1`] = `
"<h1>Admonitions</h1>
<div class="gfm-alert gfm-alert-type-note">
<p class="gfm-alert-title">Note</p>
<p>test</p>
<div class="gfm-alert gfm-alert-type-note">
<p class="gfm-alert-title">Note</p>
<p>test</p>
<div class="gfm-alert gfm-alert-type-warning">
<p class="gfm-alert-title">Warning</p>
<p>test</p>
</div>
</div>
</div>"
`;

exports[`GitHub beta blockquote-based admonitions with titles like [!NOTE] > should transform with ordered lists from issue #4 1`] = `
"<h1>Admonitions</h1>
<div class="gfm-alert gfm-alert-type-note">
<p class="gfm-alert-title">Note</p>
<p></p>
<ol>
<li>Here you go</li>
<li>Here you go again</li>
<li>Here you go one more time</li>
</ol>
</div>"
`;

exports[`GitHub beta blockquote-based admonitions with titles like [!NOTE] > should transform with title with trailing whitespaces to be trimmed 1`] = `
"<h1>Admonitions</h1>
<div class="gfm-alert gfm-alert-type-note">
<p class="gfm-alert-title">Note</p>
<p>test</p>
</div>"
`;

exports[`GitHub beta blockquote-based admonitions with titles like [!NOTE] > should transform with unordered lists from issue #4 1`] = `
"<h1>Admonitions</h1>
<div class="gfm-alert gfm-alert-type-note">
<p class="gfm-alert-title">Note</p>
<p></p>
<ul>
<li>Here you go</li>
<li>Here you go again</li>
<li>Here you go one more time</li>
</ul>
</div>"
`;

exports[`verify default behavior with plugin enabled > should transform a plain blockquote with line breaks ("  " - note the two spaces) 1`] = `
"<blockquote>
<p>Each time you increase the amount of code, your software grows exponentially more complicated.<br>
-- DHH, Getting Real</p>
</blockquote>"
`;

exports[`verify default behavior with plugin enabled > should transform a plain blockquote with line breaks (\\) 1`] = `
"<blockquote>
<p>Each time you increase the amount of code, your software grows exponentially more complicated.<br>
-- DHH, Getting Real</p>
</blockquote>"
`;

exports[`verify default behavior with plugin enabled > should transform a plain blockquote with line breaks (\\) 2x 1`] = `
"<blockquote>
<p>Each time you increase the amount of code, your software grows exponentially more complicated.<br>
-- DHH, Getting Real<br>
-- DHH, Getting Real</p>
</blockquote>"
`;
